### ISSUE TITLE
windowIconSize removed/Add a way to configure AutoSizeImage behavior

### ISSUE THREAD

**Issue by teleshoes** [taffybar/taffybar#351]
afaik, you cant scale images with css in GTK. please put this essential property back.
if i am wrong, please just provide the CSS i need to add.

if you like, you can add it as a percentage of barHeight, so that we arent actually adding any new parameters. @IvanMalison

**Comment by IvanMalison** [comment]
@teleshoes you can size icons with css by adding
padding, margin or border to the .AutoSizeImage class.

The image will always expand to fully fit the space that is allocated to them, so I suppose that the sizing is somewhat indirect. If you really want me to add the parameter back I suppose that I can, but it will substantially complicate the code.

**Comment by IvanMalison** [comment]
Of course, #350, maybe be an issue when you do this (I would appreciate if you could verify this), but I am working on fixing that right now.

**Comment by teleshoes** [comment]
i dont understand; why would i want to shrink the images but not shrink the container?
the problem is that the images are too big.

**Comment by teleshoes** [comment]
@IvanMalison perhaps i AM seeing #350, but its literally all of my images

**Comment by IvanMalison** [comment]
@teleshoes if you want to shrink the container, pad the container. If you want to shrink the bar, decrease the bar size.

**Comment by IvanMalison** [comment]
You wouldn't see #350 at startup. It happens when certain icons are added after startup. Still working on identifying the exact cause, but you would see the bar/icon size increase when opening certain windows.

**Comment by teleshoes** [comment]
yea. perhaps i am missing something, or i am confused. let me show you what i see before and after your changes:
case1, dcd014e (before auto-resize), with barHeight=50 and windowIconSize=28
![before](https://user-images.githubusercontent.com/592062/40272037-a953ef90-5b74-11e8-9182-b928f08493d8.png)


case2, 00b9308 (after windowIconSize removed), with barHeight=50
![after](https://user-images.githubusercontent.com/592062/40272050-c2645d62-5b74-11e8-9bdd-f9a55fe67b2b.png)


the squares for each workspace are almost doubled in size. how could adding padding/margin/borders to `IconContainer` or `AutoSizeImage` fix this? when i see this, i think i need to shrink the contents and have the container fit them, not the other way around.

**Comment by teleshoes** [comment]
so, what i want is to have smaller icons, without having empty space surround them, and i dont see any way to do this with GTK css, which, afaik, doesnt let you shrink things ever.

**Comment by teleshoes** [comment]
in case im crazy, i did try all combinations of padding/border/margin on all three of these classes:
`IconImage`, `IconContainer`, and `AutoSizeImage`.

-padding AutoSizeImage does in fact shrink the image, which is cute (i like AutoResizeIcon, its a nice piece), but then i just have padding instead of icon.
-adding negative margins to AutoSizeImage does what you would expect (zooms in on the image, so to speak)
-padding `IconImage` makes the squares larger.
-padding `IconContainer` does nothing, afaict

**Comment by teleshoes** [comment]
so, afaik, scaling the image to fit the container only solves my problem if i can shrink the container, which i cannot (or at least, i dont know how to).

**Comment by IvanMalison** [comment]
@teleshoes hmm. That is strange. Are you using GDK scale or something.

Can you enable the autosize logger and the StatusNotifier.Tray logger with

https://github.com/IvanMalison/dbus-hslogger
and let me know what gets logged?

**Comment by IvanMalison** [comment]
@teleshoes What I meant by shrink the container is to use the .Contents class as in the default taffybar css, but there is a problem here seems to go beyond just being able to resize the images.

The images should be sizing themselves to whatever allocation they are given.

Can you verify that this happens immediately at startup?

**Comment by IvanMalison** [comment]
@teleshoes you'll need to apply withLogServer for dbus-hslogger to work

then, after taffybar is started you can run 

```
dbus-hslogger-client -l DEBUG -p StatusNotifier.Tray
dbus-hslogger-client -l DEBUG -p System.Taffybar.Widget.Generic.AutoSizeImage
```

**Comment by IvanMalison** [comment]
to pull the appropriate logs

**Comment by IvanMalison** [comment]
> -to be clear, i dont WANT the images sizing themselves to the allocation they are given. i want them to be an explicitly smaller size.

@teleshoes 
That is fine, padding should be able to accomplish this. 

> -afaik, you cannot SHRINK things with GTK CSS, ever. you can only increase their size with min-width, padding, margins, borders, etc.

You can shrink things, when they obtain a certain size by expanding to the size of their parent by setting padding on their class. You need to have a container wrapping that object in order to do that, of course, but it is definitely possible.

> it seems to me the user can no longer shrink WorkspaceSwitcher, period.

Not true, I will post an example.

**Comment by teleshoes** [comment]
also, to be clear, AutoSizeImage is working. if i pad the container, the icon shrinks.
![padded](https://user-images.githubusercontent.com/592062/40272376-212d7e0a-5b7a-11e8-829b-1832690b901f.png)

**Comment by IvanMalison** [comment]
The first image is normal, the second image shrinks the icons, the third image shrinks workspaces

https://imgur.com/a/XXTWEhZ

**Comment by IvanMalison** [comment]
you can shrink workspaces with
```
.Workspaces {
	padding: 10px;
}
```

**Comment by teleshoes** [comment]
yes, you are correct, and i was very wrong! thanks for the lesson!
however, i dont want the entire workspaces widget padded, if you see what i mean here:
![resized](https://user-images.githubusercontent.com/592062/40272445-f20f75aa-5b7a-11e8-9b01-1e60c52bfd2d.png)
probably a matter of just setting a more specific CSS rule


ALSO, i do have what appears to be the bug in 350, but instantly at startup

**Comment by teleshoes** [comment]
to be clear, the second image in your demo is what i want. icons and workspace width are decreased, but there is no empty padding around the entire widget

**Comment by teleshoes** [comment]
whats the difference in CSS between demo#1, and demo#2?

**Comment by IvanMalison** [comment]
@teleshoes so you want the border to extend to the edge of the bar, but not the contents?

You can use the

.Contents class for that. If you don't want any left or right padding, you can do:
```
.Contents {
  padding-bottom: 2px;
  padding-top: 2px;
}
```

**Comment by IvanMalison** [comment]
@teleshoes if you pad .IconContainer instead of .AutoSizeImage you will avoid #350 for now

**Comment by IvanMalison** [comment]
@teleshoes does using .Contents work? Can I close this?

**Comment by teleshoes** [comment]
` so you want the border to extend to the edge of the bar, but not the contents?`
that does not sound like what i want, actually. what i want is to have nothing between the top of my icon and the top of my screen except my 2px orange border.
(likewise, nothing between the label of my first workspace and the left of my screen except a 2px border)

this is approximated by:
```
.Workspaces {
  padding-bottom: 16px;
}
```

but then my bar is 16px too tall, and my workspace is still a LITTLE too wide.

if there is some CSS that can give me what i had before, i cant figure it out.

**Comment by teleshoes** [comment]
like you say, i can also approximate this by:
```
.Contents {
  padding-top: 10px;
  padding-bottom: 20px;
}
```
but then i lose the bottom border.

ALSO, either way, my bar is still increased in height from what it was before.

i am not yet convinced that i can get what i had before only by applying CSS.

**Comment by teleshoes** [comment]
SHIT, the problem with the barHeight is the `Windows` widget, not the `Workspaces` widget. my apologies.

ill see if i can make this work now, though i would really MUCH prefer having `windowIconSize`.

the reason i prefer a haskell var is that i have 4 computers with different resolutions and physical monitor sizes, all running the same taffybar config. generating dynamic CSS instead of setting a haskell variable is much harder, because GTK CSS doesnt support variables in CSS (except for colors).

i CAN, in theory, calculate what the top/bottom padding should be for contents based on my desired `windowIconSize` and the `barHeight`, and then output that to a file in taffybar.hs before starting GTK.

**Comment by teleshoes** [comment]
(not sure what changed in `Windows`, but im going to revert to a simple label widget for it now.)

**Comment by teleshoes** [comment]
turns out `padding-top==padding-bottom==9px` is what i want for `bar=50px` and `desired-icon-size=28px`. ill play with it and come up with a formula, and see if i can generate the dynamic CSS.

**Comment by IvanMalison** [comment]
> SHIT, the problem with the barHeight is the Windows widget, not the Workspaces widget. my apologies.

This is what I said above.  I think you windows widget is slightly too large for you bar somehow (maybe because the label has a newline in it)


> the reason i prefer a haskell var is that i have 4 computers with different resolutions and physical monitor sizes, all running the same taffybar config. generating dynamic CSS instead of setting a haskell variable is much harder, because GTK CSS doesnt support variables in CSS (except for colors).
 
right, I have the same issue. The idea behind this change, though is that the only thing that you would have to change between computers is the bar size. I suppose that you would need to adjust the padding values as well.

How about this:

We'll add a way to globally configure autosize image to adjust the image size by some amount that is configured by a haskell integer, or even a ratio of the allocated size, so you'll be able to say I want to size this auto size image to some percentage of its actual allocation.

**Comment by IvanMalison** [comment]
> turns out padding-top==padding-bottom==9px is what i want for bar=50px and desired-icon-size=28px. ill play with it and come up with a formula, and see if i can generate the dynamic CSS.

@teleshoes Don't do that. I'd like to come up with a solution that makes sense for your use case, and is going to be easy to understand and work with for everyone.

**Comment by teleshoes** [comment]
yep, that sounds ideal!

sorry i missed when you said, probably thought you typed `Workspaces` instead of `Windows`

btw, my profiles currently have 7 variables:
```
--                 pName  barHt  wImgH  wSepW  fontP  graph  music
profileT430s = P "t430s"     50     28      3   17.0     80  15.94
profileT430  = P  "t430"     38     24      2   13.0     50  15.94
profileW520  = P  "w520"     38     24      2   13.0     50  15.63
profileNuc   = P   "nuc"     48     28      3   16.0     50  19.43
```
(each has a different DPI as well as resolution, plus one is a 65" screen that is 6' away)

wImgH will become something like image-offset-percent or something.

**Comment by teleshoes** [comment]
as for windows, though, i dont see what changed in the last few commits to cause the problem, but its not there before AutoSizeImage

**Comment by teleshoes** [comment]
also, if things work out, that variable might disappear from my profiles, and be fully defined by barHt

**Comment by IvanMalison** [comment]
@teleshoes Windows is now a button, rather than a menu bar (see #343)

**Comment by IvanMalison** [comment]
@teleshoes `constantScaleWindowIconPixbufGetter` should allow you to go back to scaling windowImages to a constant size in haskell (if you prefer doing that to auto size)

It would also be very easy to scale the icon to some proportion of the allocated size using very similar code (I'm sure you can figure out by looking at the definition of `constantScaleWindowIconPixbufGetter`

**Comment by teleshoes** [comment]
oh no, another icon rewrite! :(

sigh...just to make sure i understand this one, youve changed the icon customization back to supplying a raw pixbuf to taffybar.

if im wrong, disregard my complaints and tell me what im supposed to do instead.

if this IS what im supposed to do, im against this change for the following reasons:
1) image scaling is definitely something that taffybar should be doing, not the custom-icon loading code
2) having a fixed, predictable icon size is practical and useful, for the reasons articulated above
3) the code for fixed-icon size can also very comfortably live next to auto-resizing. perhaps you could have an enumerated scaling type, `data IconScaling = Auto | Fixed Int32 | BarPercent Double`
4) increased brittleness; to implement custom icon loading, i now have to use nuts+bolts functions that are, honestly, pretty likely to change, just looking at it statistically. (i.e.: the APIs were much higher level before, basically just datatypes)
5) if you decide, say, to do some pixbuf caching, fixed-size-icons wont be able to take advantage of it because i have manage my own low-level memory objects

@IvanMalison

**Comment by teleshoes** [comment]
also, i dont think this actually solves the problem. this only lets me control what happens when the icon is actually there.

when i use `minIcons=1` to make a placeholder, the placeholder appears to be auto-scaled, and no longer respects the fixed-icon size. i dont think i have access to that pixbuf to scale it, and afaict, getWindowIconPixbuf isnt invoked.

i just dont see any benefits to deleting the `IconInfo` datatype. it provided a nice, clean interface that was easy to understand and easy to extend. for example, to add lowlevel control, you could just add another enum value, `IconRaw Pixbuf`.

@IvanMalison

**Comment by teleshoes** [comment]
so, as a workaround, i have fixed-scale icons AND hardcoded `.contents` padding in CSS.

on a side note, i would LOVE to just open a PR fixing all this, uniting the auto-scaling and manual-scaling (and maybe relative-to-bar-scaling) with a clean interface that doesnt require custom icons to build their own pixbufs.
if you think that you might actually merge it, let me know, ill work on it soon.

**Comment by IvanMalison** [comment]
> image scaling is definitely something that taffybar should be doing, not the custom-icon loading code

@teleshoes This is a fair point, but let me explain a few things:

First of all there are two functions

`scaledWindowIconPixbufGetter`
https://github.com/taffybar/taffybar/blob/72ba688d824f49c24f96a09e91489cd38994b32d/src/System/Taffybar/Widget/Workspaces.hs#L619

`constantScaleWindowIconPixbufGetter`
https://github.com/taffybar/taffybar/blob/72ba688d824f49c24f96a09e91489cd38994b32d/src/System/Taffybar/Widget/Workspaces.hs#L624

that can be applied to any 

`WindowIconPixbufGetter`

to allow the user to select the behavior that they would like. If we were to add a third option that simply calculated the percentage of the size for you (which would obviously not be hard), this is really not significantly different than what you propose here with the custom data type:

> the code for fixed-icon size can also very comfortably live next to auto-resizing. perhaps you could have an enumerated scaling type, data IconScaling = Auto | Fixed Int32 | BarPercent Double
increased brittleness; to implement custom icon loading, i now have to use nuts+bolts functions that are, honestly, pretty likely to change, just looking at it statistically. (i.e.: the APIs were much higher level before, basically just datatypes)

I actually think that in a certain way, it is nicer that a separate icon scaling parameter because it avoids having yet another data type, and it allows users a great deal of flexibility in customizing the icon sizing behavior.

If you really strongly prefer the ADT interface, I would accept a PR that takes the parameters you are suggesting and builds the appropriate `WindowIconPixbufGetter` for the user.



> having a fixed, predictable icon size is practical and useful, for the reasons articulated above

Yes, that is the idea behind `constantScaleWindowIconPixbufGetter`



> if you decide, say, to do some pixbuf caching, fixed-size-icons wont be able to take advantage of it because i have manage my own low-level memory objects

Not really. I actually think caching needs to be very user configurable, it will be a simple matter of adding a

`cachedWindowIconPixbufGetter` function that will cache things for you. That function will be provided as part of taffybar, so no management will need to happen on your part.

Caching was actually one of the reasons I decided to do things this way.

Taffybar is now offering multiple different methods of icon lookup (EWMH, Desktop files, by window class), and it is important the it be possible to have fined grained control over what is cached and what isn't (as well as things like the rate at which the cache is invalidated, or what triggers invalidation), doing things this way will let the user, for example, cache EWMH icons but not Desktop file icons. It would be impractical to have parameters in the workspace config for all of the conceivable granularity that might be desired.

Another reason that I decided to do things this way is that I want to be able to share icon retrieval functions between widget components. I'm going to be adding window icon support to the windows icon soon, and I'm also planning on sharing the icon lookup/caching functionality with the tray widget system. The current interface is generic enough that it allows code sharing across all of these modules.

**Comment by IvanMalison** [comment]
> when i use minIcons=1 to make a placeholder, the placeholder appears to be auto-scaled, and no longer respects the fixed-icon size. i dont think i have access to that pixbuf to scale it, and afaict, getWindowIconPixbuf isnt invoked.

Yeah, I suppose this is a problem, but it's not really related to the issue of scaling, or the use of raw pixbufs imo. I think what needs to happen here is

WindowIconPixbufGetter should be of type 

```
type WindowIconPixbufGetter =
  Int32 -> Maybe WindowData -> TaffyIO (Maybe Gdk.Pixbuf)
```

That way a pixbuf can be provided when you want to draw an icon for no pixbuf. How does that sound?

**Comment by IvanMalison** [comment]
> on a side note, i would LOVE to just open a PR fixing all this, uniting the auto-scaling and manual-scaling

In summary, you seem to be proposing two separate things here:

The first is to add a completely new data type:

```
data IconScaling = Auto | Fixed Int32 | BarPercent Double
```

which would affect how window icon pixbufs are scaled. I'm still not convinced that this is significantly different than simply having `WindowIconPixbufGetter -> WindowIconPixbufGetter` of the same name that do the scaling for you. This avoids yet another configuration parameter.

Still, i am more open to that than:
> i just dont see any benefits to deleting the IconInfo datatype.

Personally I see almost no benefit to the IconInfo datatype:

First of all, it was inflexible. Now that taffybar is offering all of these different sorts of icon retrieval mechanisms

`getWindowIconPixbufFromEWMH`
`getWindowIconPixbufFromClass`
`getWindowIconPixbufFromDesktopEntry`

we absolutely need a way for the user to declare which of these approaches should take priority over the other. By having this raw monadic function interface, we can do this easily. Taffybar even provides the `<|||>` operator for combining WindowIconPixbufGetter types in an intuitive way.

The other advantage this confers is that if someone comes up with an entirely new way of retrieving the window icon for their windows they can easily implement a new function to do so. Without an interface like this, this is not possible. I suppose you could do something like you suggested with a constructor `IconRaw Pixbuf`, but that really is, ultimately, the same thing.

Basically, this boils down to for me, how is

```
IIColor myColorValue
```

Any different from

```
pixBufFromColor mySize myColorValue
```

except that the latter allows the user to size the pixbuf (which we've agreed is a good thing).

I don't see the IconInfo data type as adding anything except for constraints, basically.

**Comment by IvanMalison** [comment]
@teleshoes I don't want to come off as combative. I really appreciate your feedback, and I want to understand and cater to your needs.

I'm pretty convinced that the interface that has been exposed is the way to go, but if you feel really strange about it, I'm open to adding a parameter to control scaling.

My feeling is just that functions offer more control, and Haskell is a VERY function oriented language that makes it really easy to manipulate functions as values.

I don't see much of a difference between passing a particular function value and passing values that ultimately select between a fixed number of behavior, except that the former is much more flexible.

That's where I'm coming from. If you feel like there is something specific that you are unable to accomplish (without very verbose configuration) with the current interface, then let's look at what that is and figure out how to make it possible.

**Comment by teleshoes** [comment]
oh yea, no worries, i see where youre coming from completely. i just disagree on some of the finer points.

also, if I'M coming across as hostile, its absolutely unintentional. i am a little overwhelmed with how often icons keep changing, though; i find that i need to rewrite my custom icon code pretty much from scratch every coupla weeks. things ARE improving enormously, though. i suppose churn is just an unpleasant side effect of progress.

> I don't see much of a difference between passing a particular function value and passing values that ultimately select between a fixed number of behavior, except that the former is much more flexible

so when deciding how high-level to make a given API, theres always a tradeoff between flexibility and ease-of-use. i think in this particular case, constructing low-level graphics objects in order to have fixed-size icons is where id draw that line, but a large part of the problem is that my code doesnt take advantage of all the utility fcts you wrote to support it yet.

> we absolutely need a way for the user to declare which of these approaches should take priority over the other

yea thats mostly what my custom icon code is about. my current order is:
1) CustomOverride (if title/class match 9 or 10 specific values, use that, regardless of EWMH icons)
2) EWMH (if ewmh icon is present, use that)
3) CustomNormal (select icons from a custom dir based on title/class, for when there is no EWMH icon)
4) Placeholder (a single image with a question mark, to differentiate between an empty workspace and windows with absolutely no icon)

this is probably a little crazy, but i live and die by it, and its not too unreasonable anyway.

> ..they can easily implement a new function to do so. Without an interface like this, this is not possible

heh, except thats exactly what ive had since i first added icons to taffybar; a fully custom icon-selector function, based entirely on title/class.

> ...Maybe WindowData...That way a pixbuf can be provided when you want to draw an icon for no pixbuf. How does that sound?

yea that completely solves that particular problem, i think. i suppose i would call it getWorkspacePixbufNoWindow or something, and it would have the form `f Nothing = Just emptyPixbuf; f (Just _) = Nothing`

> I'm open to adding a parameter to control scaling.

so i still think it would be nice to separate out image selection from image scaling, but i cant tell for sure yet.
after i rewrite my custom icon selectors in terms of getWsPixbufs and you change WindowData=>Maybe WindowData, i might agree with you that it makes sense to treat scaling and image selection as one process

and as always, thanks for taking the time and energy to keep making taffybar better and better

**Comment by IvanMalison** [comment]
> i find that i need to rewrite my custom icon code pretty much from scratch every coupla weeks.

Yeah, sorry about that. I hope that we're converging to an approach that will remain fixed for the foreseeable future.

> constructing low-level graphics objects in order to have fixed-size icons is where id draw that line

I agree that the user should not have to think about pixbufs to get fixed size icons, but just so we're clear `constantScaleWindowIconPixbufGetter` makes it so that that isn't necessary, right?

> heh, except thats exactly what ive had since i first added icons to taffybar; a fully custom icon-selector function, based entirely on title/class.

Well before, you had to choose from one of the provided options (ewmh, color or a filepath). I have now added other ways to load icons (from desktop entries, for example), that would not have been possible with the old interface. By saying "hey you can give us any pixbuf" we avoid backwards incompatibility, and flexibility issues there. I really don't think much is lost either, because the translations are really straightforward.

IIColor is just

`pixbufFromColor`

IIFilePath is just

`pixbufFromFilePath`

> so i still think it would be nice to separate out image selection from image scaling, but i cant tell for sure yet.

I just realized that the nice thing about having a separate parameter for scaling, is that it allows us to avoid changing the signature of `WindowIconPixbufGetter`, since we can still apply the scaling function to the transparent pixbuf that is generated when there is no window.

Sort of kills two birds with one stone. The signature would be

```
Int32 -> Pixbuf -> TaffyIO Pixbuf
```

and of course we would provide a bunch of default values like:

`scaleToPercentages`, `scaleToFixedSize`, `scaleToFill`

Does that sound like a workable solution?

**Comment by teleshoes** [comment]
yea i think that will work, though you might want to add another setting to the record for get-empty-pixbuf to keep the flexibility to do it the other way. `getEmptyWindowPixbuf = Int32 -> TaffyIO (Maybe Pixbuf)`

so, while i was rewriting to build pixbufs, i considered preloading the pixbufs for my custom icons and just reusing them. it occurs to me that if you do scaling separately, then this optimization will be pointless. if its efficient, could the scaling be skipped if its unnecessary? (with `pixbufGetHeight`?)

one more thing; i am ignoring the size argument being passed in for all my `WindowIconPixbufGetter`s and using the fixed value from my config.
it feels a bit weird to do so; am i missing something?

**Comment by IvanMalison** [comment]
> one more thing; i am ignoring the size argument being passed in for all my WindowIconPixbufGetters and using the fixed value from my config.
> it feels a bit weird to do so; am i missing something?

No, that is what you should do if you want constant size scaling. Of course, if you want percentage scaling (or auto size scaling, the pixbuf should be scaled to this size). It's just away of letting you set the image size based on the amount of available space.

>  it occurs to me that if you do scaling separately, then this optimization will be pointless. if its efficient, could the scaling be skipped if its unnecessary? (with pixbufGetHeight?)

Yes, this was actually one of the thoughts that originally led me to decide to combine scaling and pixbuf selection.

Makes me think that that might be the way to go, still.
